### ISSUE TITLE
Gradients through MultivariateNormal distribution log_prob behaving strangely

### ISSUE THREAD

**Issue by fleeb24** [pytorch/pytorch#8108]
I would like to use MultivariateNormal distributions to compute the log probs of some samples and then differentiate wrt to the distribution mean and covariance matrix. During testing, I encountered a strange inconsistency when trying to compute the gradients wrt the covariance matrix.

Here's a little test script to test how log_prob behaves, please let me know if you get different results than me

```
import torch
from torch.autograd import grad
import torch.distributions as distrib
def get_n():
    mu = torch.Tensor([1,2])#torch.randn(2)
    mu.requires_grad = True
    sigma = torch.Tensor([2,3])#torch.rand(2) * 10 + 1
    sigma.requires_grad = True

    n = distrib.MultivariateNormal(loc=mu, covariance_matrix=torch.diag(sigma))
    return n
def check(fn):
    try:
        fn()
    except Exception as e:
        return 'failed - {}'.format(e)
    return 'passed'
def test1():
    n = get_n()
    s = torch.ones(2)
    grad(n.log_prob(s).sum(), n.covariance_matrix)#, retain_graph=True))
def test2():
    n = get_n()
    grad(n.log_prob(n.sample()).sum(), n.covariance_matrix)
def test3():
    n = get_n()
    s = torch.ones(2)
    grad(n.log_prob(s).sum(), n.covariance_matrix, retain_graph=True)
    grad(n.log_prob(n.sample()).sum(), n.covariance_matrix, retain_graph=True)
def test4():
    n = get_n()
    grad(n.log_prob(n.sample()).sum(), n.covariance_matrix, retain_graph=True)
print(check(test1)) # this test passes for me
print(check(test2)) # this test fails (One of the differentiated Tensors appears to not have been used in the graph)
print(check(test3)) # this test passes
print(check(test4)) # this test fails (One of the differentiated Tensors appears to not have been used in the graph)

```
## System Info
Ubuntu 14.04 - Anaconda 5.x - Python 3.6 (I don't think this is system dependent)

Thanks for the help!

**Comment by vishwakftw** [comment]
I am not able to reproduce this on master. Could you tell which version of PyTorch you are using?

**Comment by vishwakftw** [comment]
Unfortunately, I am not able to understand why the error happened. Seems to have been fixed somehow :-) .

**Comment by vishwakftw** [comment]
Can this issue be closed?

cc: @apaszke
